specs/transport/car/carv2/index.md

```diff
@@ -0,0 +1,177 @@
+
+The CARv1 data payload ends after the number of bytes indicated by the "Data length" field in the header. This is important as the CARv1 parser has no means to determine the end of the container other than encountering the end of a byte stream. Therefore, a CARv2 decoder may defer to a CARv1 decoder to load the `CID:Bytes` sequence payload and "roots" array, but must be able to deliver it a byte stream that strictly begins at "Data offset" and ends at "Data length".
+
+The CARv1 data payload must include a valid CARv1 header, including a roots array and a `version` field with a value strictly of `1`.
+
+Refer to the [CARv1 Specification](../carv1/index.md) for details on the CARv1 format.
+
+### Index payload
+
+The CARv2 index payload follows the CARv1 data payload, but may be offset by padding or additional data as dictated by the characteristics field. The index payload begins at "Index offset", which _must_ be after the end of the CARv2 data payload, and continues until the end of the CARv2 byte stream (its length is not encoded in the header).
+
+An "Index offset" value of `0` in the CARv2 header indicates that there is no index in this CARv2 container and no attempt should be made to read it.
+
+### Index format
+
+The CARv2 index is a flexible container format itself, allowing for different index layouts depending on suitability for a particular application or set of data - generation speed, usage performance, size, etc. Supported index formats will be detailed in this specification, below.
+
+Index data, once read from the CARv2 container, provides a mapping of hash digest bytes to block location in byte offset from the beginning of the CARv2 container. The index only uses the hash digest&mdash;it does not use the full bytes of a CID, nor does it use any of the multihash prefix bytes.
+
+The first byte(s) of a CARv2 index are a varint that indicates the index format type. The remaining bytes follow the encoding rules of that index format type.
+
+As the index only contains the hash digest bytes, other details contained within the block's CID and the length of the block's bytes must be derived by inspecting the initial bytes of the block entry within the data payload.
+
+**TODO:** How are we handling identity hash CIDs?
+* Do we want to state that identity multihash CIDs are explicitly excluded from indexes? 
+* Can we defer an assumption that a block store interface would return the bytes without having to look them up when it sees an identity CID? 
+* What are the implications of having an index that doesn't contain all of the blocks within the CAR?
+* Or do we want to explicitly exclude index CIDs entirely from the CARv2 format since they are arguably a waste of space?
```

Comment:
I think it makes sense to say that one shouldn't use CAR files for identity hash CIDs. In general, it just doesn't make sense.

I also think it's reasonable for the index to refuse to index those; that allows the index to know that hashes are relatively small, for example. If we had an API that takes a CID or Hash and gives you an `io.Reader` of the data in Go, the logic could be like:

1) Is it an identity hash? If so, simply return a reader of the identity hash itself
2) Otherwise, look it up in the index, and return a reader of the block slice in the CARv1

Something worth thinking about is the ability to store entire DAGs in a CAR file. If we forbid storing identity hash CIDs as blocks in this spec, does that mean it's impossible to store a complete DAG in a CARv2 when it has an identity hash CID link? All the data would be there, since the link would contain the whole block, but strictly speaking some CIDs would be missing from the CARv1.

Reply:
I agree, there’s really no utility in including them, they appear whole in any block that references them, that’s enough.

Reply:
OK, so next question - do we take this opportunity to explicitly rule them out of the CARv1 contained within the CARv2 here? Or is that doing narrowing down the existing spec a bit too much? This will end up being a question of determinism for cases where it matters (like Filecoin deals). Maybe this is a "characteristic"?